File: src/main/java/com/uiptv/service/LogoResolverService.java
package com.uiptv.service;

import com.uiptv.util.HttpUtil;
import org.json.JSONArray;
import org.json.JSONObject;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

import static com.uiptv.util.StringUtils.isBlank;
import static com.uiptv.util.StringUtils.isNotBlank;

public class LogoResolverService {
    private static final String CHANNELS_CATALOG_URL = "https://iptv-org.github.io/api/channels.json";
    private static final String LOGOS_CATALOG_URL = "https://raw.githubusercontent.com/iptv-org/database/master/data/logos.csv";
    private static final long CATALOG_REFRESH_MS = 24L * 60L * 60L * 1000L;
    private static final LogoResolverService INSTANCE = new LogoResolverService();
    private static final Set<String> NOISE_TOKENS = new HashSet<>(Arrays.asList(
            "uhd", "fhd", "hd", "sd", "hq", "4k", "8k",
            "hevc", "h264", "h265", "x264", "x265",
            "tv", "channel", "live", "official",
            "plus", "intl", "international"
    ));

    private final ConcurrentHashMap<String, String> localCache = new ConcurrentHashMap<>();
    private final ConcurrentHashMap<String, String> catalog = new ConcurrentHashMap<>();
    private final File localCacheFile;
    private volatile long lastCatalogRefreshAt = 0L;
    private volatile boolean refreshInProgress = false;

    private LogoResolverService() {
        this.localCacheFile = new File(System.getProperty("java.io.tmpdir"), "uiptv-logo-cache.json");
        loadLocalCache();
    }

    public static LogoResolverService getInstance() {
        return INSTANCE;
    }

    public String resolve(String channelName, String providerLogo, String countryCode) {
        if (isNotBlank(providerLogo)) {
            return providerLogo;
        }
        if (isBlank(channelName)) {
            return "";
        }

        String key = makeLookupKey(channelName);
        if (isBlank(key)) {
            return "";
        }

        String cached = localCache.get(key);
        if (isNotBlank(cached)) {
            return cached;
        }

        triggerCatalogRefreshAsync();
        String resolved = catalog.get(key);
        if (isBlank(resolved)) {
            resolved = catalog.get(makeLookupKey(stripCommonSuffixes(channelName)));
        }
        if (isBlank(resolved)) {
            resolved = resolveByVariants(channelName);
        }
        if (isBlank(resolved)) {
            resolved = resolveByTokenSubset(channelName);
        }

        if (isNotBlank(resolved)) {
            localCache.put(key, resolved);
            persistLocalCache();
            return resolved;
        }
        return "";
    }

    private synchronized void ensureCatalogLoaded() {
        long now = System.currentTimeMillis();
        if (!catalog.isEmpty() && now - lastCatalogRefreshAt < CATALOG_REFRESH_MS) {
            return;
        }
        if (refreshInProgress) {
            return;
        }
        refreshInProgress = true;
        try {
            HttpUtil.HttpResult response = HttpUtil.sendRequest(CHANNELS_CATALOG_URL, null, "GET");
            if (response.statusCode() != HttpUtil.STATUS_OK) {
                return;
            }
            String json = response.body();
            JSONArray channels = new JSONArray(json);
            Map<String, String> logoByChannelId = loadLogosByChannelId();

            ConcurrentHashMap<String, String> fresh = new ConcurrentHashMap<>();
            for (int i = 0; i < channels.length(); i++) {
                JSONObject item = channels.getJSONObject(i);
                String channelId = item.optString("id", "");
                String logo = item.optString("logo", "");
                if (isBlank(logo) && isNotBlank(channelId)) {
                    logo = logoByChannelId.getOrDefault(channelId, "");
                }
                if (isBlank(logo)) {
                    continue;
                }

                String name = item.optString("name", "");
                addAlias(fresh, name, logo);

                JSONArray altNames = item.optJSONArray("alt_names");
                if (altNames != null) {
                    for (int j = 0; j < altNames.length(); j++) {
                        addAlias(fresh, altNames.optString(j, ""), logo);
                    }
                }
            }

            if (!fresh.isEmpty()) {
                catalog.clear();
                catalog.putAll(fresh);
                lastCatalogRefreshAt = now;
            }
        } catch (Exception ignored) {
            // keep using local/provider fallback only
        } finally {
            refreshInProgress = false;
        }
    }

    private Map<String, String> loadLogosByChannelId() {
        ConcurrentHashMap<String, String> logoById = new ConcurrentHashMap<>();
        try {
            HttpUtil.HttpResult response = HttpUtil.sendRequest(LOGOS_CATALOG_URL, null, "GET");
            if (response.statusCode() != HttpUtil.STATUS_OK) {
                return logoById;
            }
            String csv = response.body();
            String[] lines = csv.split("\\r?\\n");
            for (int i = 1; i < lines.length; i++) {
                String line = lines[i];
                if (isBlank(line)) continue;
                List<String> cells = parseCsvLine(line);
                if (cells.size() < 7) continue;
                String channel = cells.get(0);
                String logoUrl = cells.get(6);
                if (isBlank(channel) || isBlank(logoUrl)) continue;
                logoById.putIfAbsent(channel, logoUrl);
            }
        } catch (Exception ignored) {
            // best-effort only
        }
        return logoById;
    }

    private List<String> parseCsvLine(String line) {
        List<String> cells = new ArrayList<>();
        StringBuilder current = new StringBuilder();
        boolean inQuotes = false;
        for (int i = 0; i < line.length(); i++) {
            char c = line.charAt(i);
            if (c == '"') {
                if (inQuotes && i + 1 < line.length() && line.charAt(i + 1) == '"') {
                    current.append('"');
                    i++;
                } else {
                    inQuotes = !inQuotes;
                }
            } else if (c == ',' && !inQuotes) {
                cells.add(current.toString());
                current.setLength(0);
            } else {
                current.append(c);
            }
        }
        cells.add(current.toString());
        return cells;
    }

    private void triggerCatalogRefreshAsync() {
        long now = System.currentTimeMillis();
        if ((!catalog.isEmpty() && now - lastCatalogRefreshAt < CATALOG_REFRESH_MS) || refreshInProgress) {
            return;
        }
        Thread loader = new Thread(this::ensureCatalogLoaded, "uiptv-logo-catalog-refresh");
        loader.setDaemon(true);
        loader.start();
    }

    private void addAlias(Map<String, String> target, String alias, String logo) {
        String key = makeLookupKey(alias);
        if (isNotBlank(key) && isNotBlank(logo)) {
            target.putIfAbsent(key, logo);
        }
    }

    private String stripCommonSuffixes(String value) {
        if (value == null) {
            return "";
        }
        String cleaned = value.replaceAll("(?i)\\b(\\+1|\\+2|\\+3|\\+4)\\b", " ");
        cleaned = cleaned.replaceAll("(?i)\\b(uhd|fhd|hd|sd|hq|4k|8k|hevc|h264|h265|x264|x265|tv|channel|live|official|intl|international)\\b", " ");
        return cleaned.trim();
    }

    private String makeLookupKey(String value) {
        if (isBlank(value)) {
            return "";
        }
        return value.toLowerCase()
                .replace('&', ' ')
                .replace('+', ' ')
                .replaceAll("[^a-z0-9 ]", " ")
                .replaceAll("\\s+", " ")
                .trim();
    }

    private String resolveByVariants(String originalName) {
        List<String> variants = buildNameVariants(originalName);
        for (String variant : variants) {
            String hit = catalog.get(variant);
            if (isNotBlank(hit)) {
                return hit;
            }
        }
        return "";
    }

    private List<String> buildNameVariants(String originalName) {
        String base = makeLookupKey(originalName);
        String stripped = makeLookupKey(stripCommonSuffixes(originalName));

        Set<String> variants = new HashSet<>();
        if (isNotBlank(base)) variants.add(base);
        if (isNotBlank(stripped)) variants.add(stripped);

        List<String> tokenList = new ArrayList<>();
        for (String token : stripped.split(" ")) {
            if (isBlank(token)) continue;
            if (NOISE_TOKENS.contains(token)) continue;
            tokenList.add(token);
        }
        if (!tokenList.isEmpty()) {
            variants.add(String.join(" ", tokenList));
            // progressively remove trailing token(s), useful for "set max hd", "sony max 2 hd"
            for (int i = tokenList.size() - 1; i >= 1; i--) {
                variants.add(String.join(" ", tokenList.subList(0, i)));
            }
        }

        return new ArrayList<>(variants);
    }

    private String resolveByTokenSubset(String originalName) {
        Set<String> tokens = new HashSet<>();
        for (String token : makeLookupKey(stripCommonSuffixes(originalName)).split(" ")) {
            if (isBlank(token) || NOISE_TOKENS.contains(token)) continue;
            tokens.add(token);
        }
        if (tokens.isEmpty()) {
            return "";
        }

        // Best effort fuzzy: choose first catalog name containing all significant tokens.
        for (Map.Entry<String, String> entry : catalog.entrySet()) {
            String catalogKey = entry.getKey();
            boolean allMatch = true;
            for (String token : tokens) {
                if (!catalogKey.contains(token)) {
                    allMatch = false;
                    break;
                }
            }
            if (allMatch && isNotBlank(entry.getValue())) {
                return entry.getValue();
            }
        }
        return "";
    }

    private void loadLocalCache() {
        if (!localCacheFile.exists()) {
            return;
        }
        try (FileInputStream in = new FileInputStream(localCacheFile)) {
            JSONObject root = new JSONObject(new String(in.readAllBytes(), StandardCharsets.UTF_8));
            for (String key : root.keySet()) {
                String value = root.optString(key, "");
                if (isNotBlank(value)) {
                    localCache.put(key, value);
                }
            }
        } catch (Exception ignored) {
            // ignore corrupt cache
        }
    }

    private synchronized void persistLocalCache() {
        try (FileOutputStream out = new FileOutputStream(localCacheFile)) {
            out.write(new JSONObject(localCache).toString().getBytes(StandardCharsets.UTF_8));
        } catch (Exception ignored) {
            // best-effort cache only
        }
    }
}

File: server/src/main/kotlin/com/uiptv/service/LogoResolverService.kt
package com.uiptv.service

import com.uiptv.util.HttpUtil
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.nio.charset.StandardCharsets
import java.util.concurrent.ConcurrentHashMap
import com.uiptv.util.StringUtils.isBlank
import com.uiptv.util.StringUtils.isNotBlank
import org.json.JSONArray
import org.json.JSONObject

object LogoResolverService {
    private const val CHANNELS_CATALOG_URL = "https://iptv-org.github.io/api/channels.json"
    private const val LOGOS_CATALOG_URL = "https://raw.githubusercontent.com/iptv-org/database/master/data/logos.csv"
    private const val CATALOG_REFRESH_MS = 24L * 60L * 60L * 1000L
    private val NOISE_TOKENS = setOf(
        "uhd", "fhd", "hd", "sd", "hq", "4k", "8k",
        "hevc", "h264", "h265", "x264", "x265",
        "tv", "channel", "live", "official",
        "plus", "intl", "international"
    )

    private val localCache = ConcurrentHashMap<String, String>()
    private val catalog = ConcurrentHashMap<String, String>()
    private val localCacheFile = File(System.getProperty("java.io.tmpdir"), "uiptv-logo-cache.json")
    @Volatile
    private var lastCatalogRefreshAt = 0L
    @Volatile
    private var refreshInProgress = false

    init {
        loadLocalCache()
    }

    @JvmStatic
    fun getInstance(): LogoResolverService = this
    fun resolve(channelName: String?, providerLogo: String?): String {
        if (isNotBlank(providerLogo)) return providerLogo.orEmpty()
        if (isBlank(channelName)) return ""
        val key = makeLookupKey(channelName)
        if (isBlank(key)) return ""
        val cached = localCache[key]
        if (isNotBlank(cached)) return cached.orEmpty()
        val resolved = resolveFromCatalog(channelName!!, key)
        return cacheResolvedLogo(key, resolved)
    }

    @Synchronized
    private fun ensureCatalogLoaded() {
        val now = System.currentTimeMillis()
        if (isCatalogFresh(now) || refreshInProgress) return
        refreshInProgress = true
        try {
            val response = fetchChannelsCatalog()
            if (response.statusCode != HttpUtil.STATUS_OK) return
            val logoByChannelId = loadLogosByChannelId()
            val fresh = buildCatalogEntries(JSONArray(response.body), logoByChannelId)
            if (fresh.isNotEmpty()) {
                catalog.clear()
                catalog.putAll(fresh)
                lastCatalogRefreshAt = now
            }
        } catch (_: Exception) {
        } finally {
            refreshInProgress = false
        }
    }

    private fun isCatalogFresh(now: Long): Boolean = catalog.isNotEmpty() && now - lastCatalogRefreshAt < CATALOG_REFRESH_MS

    private fun fetchChannelsCatalog(): HttpUtil.HttpResult = HttpUtil.sendRequest(CHANNELS_CATALOG_URL, null, "GET")

    private fun buildCatalogEntries(channels: JSONArray, logoByChannelId: Map<String, String>): ConcurrentHashMap<String, String> {
        val fresh = ConcurrentHashMap<String, String>()
        for (index in 0 until channels.length()) {
            channels.optJSONObject(index)?.let { addCatalogAliases(fresh, it, logoByChannelId) }
        }
        return fresh
    }

    private fun addCatalogAliases(fresh: ConcurrentHashMap<String, String>, item: JSONObject, logoByChannelId: Map<String, String>) {
        val logo = resolveCatalogLogo(item, logoByChannelId)
        if (isBlank(logo)) return
        addAlias(fresh, item.optString("name", ""), logo)
        val altNames = item.optJSONArray("alt_names") ?: return
        for (index in 0 until altNames.length()) {
            addAlias(fresh, altNames.optString(index, ""), logo)
        }
    }

    private fun resolveCatalogLogo(item: JSONObject, logoByChannelId: Map<String, String>): String {
        val channelId = item.optString("id", "")
        val logo = item.optString("logo", "")
        return if (isBlank(logo) && isNotBlank(channelId)) logoByChannelId[channelId].orEmpty() else logo
    }

    private fun resolveFromCatalog(channelName: String, key: String): String {
        triggerCatalogRefreshAsync()
        var resolved = catalog[key]
        if (isBlank(resolved)) resolved = catalog[makeLookupKey(stripCommonSuffixes(channelName))]
        if (isBlank(resolved)) resolved = resolveByVariants(channelName)
        if (isBlank(resolved)) resolved = resolveByTokenSubset(channelName)
        return resolved.orEmpty()
    }

    private fun cacheResolvedLogo(key: String, resolved: String?): String {
        if (isBlank(resolved)) return ""
        localCache[key] = resolved!!
        persistLocalCache()
        return resolved
    }

    private fun loadLogosByChannelId(): Map<String, String> {
        val logoById = ConcurrentHashMap<String, String>()
        try {
            val response = HttpUtil.sendRequest(LOGOS_CATALOG_URL, null, "GET")
            if (response.statusCode != HttpUtil.STATUS_OK) return logoById
            val lines = response.body.split(Regex("\\r?\\n"))
            for (index in 1 until lines.size) {
                val line = lines[index]
                if (isBlank(line)) continue
                val cells = parseCsvLine(line)
                if (cells.size < 7) continue
                val channel = cells[0]
                val logoUrl = cells[6]
                if (isBlank(channel) || isBlank(logoUrl)) continue
                logoById.putIfAbsent(channel, logoUrl)
            }
        } catch (_: Exception) {
        }
        return logoById
    }

    private fun parseCsvLine(line: String): List<String> {
        val cells = ArrayList<String>()
        val current = StringBuilder()
        var inQuotes = false
        var index = 0
        while (index < line.length) {
            val c = line[index]
            if (c == '"') {
                if (inQuotes && index + 1 < line.length && line[index + 1] == '"') {
                    current.append('"')
                    index++
                } else {
                    inQuotes = !inQuotes
                }
            } else if (c == ',' && !inQuotes) {
                cells.add(current.toString())
                current.setLength(0)
            } else {
                current.append(c)
            }
            index++
        }
        cells.add(current.toString())
        return cells
    }

    private fun triggerCatalogRefreshAsync() {
        val now = System.currentTimeMillis()
        if ((catalog.isNotEmpty() && now - lastCatalogRefreshAt < CATALOG_REFRESH_MS) || refreshInProgress) return
        Thread(this::ensureCatalogLoaded, "uiptv-logo-catalog-refresh").apply {
            isDaemon = true
            start()
        }
    }

    private fun addAlias(target: MutableMap<String, String>, alias: String?, logo: String?) {
        val key = makeLookupKey(alias)
        if (isNotBlank(key) && isNotBlank(logo)) {
            target.putIfAbsent(key, logo!!)
        }
    }

    private fun stripCommonSuffixes(value: String?): String {
        if (value == null) return ""
        var cleaned = value.replace(Regex("(?i)\\b(\\+1|\\+2|\\+3|\\+4)\\b"), " ")
        cleaned = cleaned.replace(Regex("(?i)\\b(uhd|fhd|hd|sd|hq|4k|8k|hevc|h264|h265|x264|x265|tv|channel|live|official|intl|international)\\b"), " ")
        return cleaned.trim()
    }

    private fun makeLookupKey(value: String?): String {
        if (isBlank(value)) return ""
        return value!!.lowercase()
            .replace('&', ' ')
            .replace('+', ' ')
            .replace(Regex("[^a-z0-9 ]"), " ")
            .replace(Regex("\\s+"), " ")
            .trim()
    }

    private fun resolveByVariants(originalName: String): String {
        buildNameVariants(originalName).forEach { variant ->
            val hit = catalog[variant]
            if (isNotBlank(hit)) return hit.orEmpty()
        }
        return ""
    }

    private fun buildNameVariants(originalName: String): List<String> {
        val base = makeLookupKey(originalName)
        val stripped = makeLookupKey(stripCommonSuffixes(originalName))
        val variants = LinkedHashSet<String>()
        if (isNotBlank(base)) variants.add(base)
        if (isNotBlank(stripped)) variants.add(stripped)
        val tokenList = ArrayList<String>()
        stripped.split(" ").forEach { token ->
            if (isBlank(token) || NOISE_TOKENS.contains(token)) return@forEach
            tokenList.add(token)
        }
        if (tokenList.isNotEmpty()) {
            variants.add(tokenList.joinToString(" "))
            for (index in tokenList.size - 1 downTo 1) {
                variants.add(tokenList.subList(0, index).joinToString(" "))
            }
        }
        return ArrayList(variants)
    }

    private fun resolveByTokenSubset(originalName: String): String {
        val tokens = LinkedHashSet<String>()
        makeLookupKey(stripCommonSuffixes(originalName)).split(" ").forEach { token ->
            if (isBlank(token) || NOISE_TOKENS.contains(token)) return@forEach
            tokens.add(token)
        }
        if (tokens.isEmpty()) return ""
        for ((catalogKey, value) in catalog.entries) {
            val allMatch = tokens.all { catalogKey.contains(it) }
            if (allMatch && isNotBlank(value)) return value
        }
        return ""
    }

    private fun loadLocalCache() {
        if (!localCacheFile.exists()) return
        try {
            FileInputStream(localCacheFile).use { input ->
                val root = JSONObject(String(input.readAllBytes(), StandardCharsets.UTF_8))
                root.keys().forEach { key ->
                    val value = root.optString(key, "")
                    if (isNotBlank(value)) localCache[key] = value
                }
            }
        } catch (_: Exception) {
        }
    }

    @Synchronized
    private fun persistLocalCache() {
        try {
            FileOutputStream(localCacheFile).use { output ->
                output.write(JSONObject(localCache.toMap()).toString().toByteArray(StandardCharsets.UTF_8))
            }
        } catch (_: Exception) {
        }
    }
}
